dataframe/series.go

```diff
@@ -0,0 +1,423 @@
+		}
+		return val, true, nil
+	}
+	return starlark.None, false, fmt.Errorf("not found: %v", keyVal)
+}
+
+func (s *Series) stringify() string {
+	// Calculate how wide the index column needs to be
+	indexWidth := 0
+	if len(s.index) == 0 {
+		indexWidth = len(fmt.Sprintf("%d", s.len()-1))
+	} else {
+		for _, elem := range s.index {
+			w := len(elem)
+			if w > indexWidth {
+				indexWidth = w
+			}
+		}
+	}
+
+	// Calculate how wide the data column needs to be
+	colWidth := 0
+	for _, elem := range s.stringValues() {
+		w := len(elem)
+		if w > colWidth {
+			colWidth = w
+		}
+	}
+
+	// Final line shows the (optional) name and dtype
+	epilogue := fmt.Sprintf("dtype: %s", s.dtype)
+	if s.dtype == "" {
+		epilogue = fmt.Sprintf("dtype: int64")
+	}
+	if s.name != "" {
+		epilogue = fmt.Sprintf("Name: %s, %s", s.name, epilogue)
+	}
+
+	// Determine how to format each line, based upon the column width
+	padding := "    "
+	var tmpl string
+	if len(s.index) == 0 {
+		// Result looks like '%-2d    %6s'
+		tmpl = fmt.Sprintf("%%-%dd%s%%%ds", indexWidth, padding, colWidth)
+	} else {
+		// Result looks like '%-4s    %6s'
+		tmpl = fmt.Sprintf("%%-%ds%s%%%ds", indexWidth, padding, colWidth)
+	}
+
+	// Render each value in the series
+	render := make([]string, 0, s.len()+1)
+	for i, elem := range s.stringValues() {
+		line := ""
+		if len(s.index) == 0 {
+			line = fmt.Sprintf(tmpl, i, elem)
+		} else {
+			line = fmt.Sprintf(tmpl, s.index[i], elem)
+		}
+		render = append(render, line)
+	}
+
+	// Combine the lines together
+	render = append(render, epilogue)
+	return strings.Join(render, "\n")
+}
+
+// values returns a slice of some go native type
+func (s *Series) values() []interface{} {
+	if s.which == TypeInt {
+		result := make([]interface{}, len(s.valInts))
+		for i, elem := range s.valInts {
+			result[i] = elem
+		}
+		return result
+	} else if s.which == TypeFloat {
+		result := make([]interface{}, len(s.valFloats))
+		for i, elem := range s.valFloats {
+			result[i] = elem
+		}
+		return result
+	}
+	result := make([]interface{}, len(s.valObjs))
+	for i, elem := range s.valObjs {
+		result[i] = elem
+	}
+	return result
+}
+
+// stringValues returns a slice of the stringified values, fit for printing
+func (s *Series) stringValues() []string {
+	if s.which == TypeInt {
+		result := make([]string, len(s.valInts))
+		for i, elem := range s.valInts {
+			result[i] = strconv.Itoa(elem)
+		}
+		return result
+	} else if s.which == TypeFloat {
+		result := make([]string, len(s.valFloats))
+		for i, elem := range s.valFloats {
+			result[i] = fmt.Sprintf("%1.1f", elem)
+		}
+		return result
+	}
+	return s.valObjs
+}
+
+// len returns the number of values
+func (s *Series) len() int {
+	if s.which == TypeInt {
+		return len(s.valInts)
+	} else if s.which == TypeFloat {
+		return len(s.valFloats)
+	}
+	return len(s.valObjs)
+}
+
+// takeFirst returns a series containing just the first 'n' values
+func (s *Series) takeFirst(n int) Series {
```

Comment:
this func is never hit by tests. Would love an added case that works it, even just the happy path

Reply:
Removed, it isn't needed yet.